Make it easy to add system-wide custom channels.

aszlig opened the pull request:
This is very useful if you want to distribute channels (and thus expressions as well) in a similar fashion to Debians APT sources (or PPAs or whatnot).

So, for example if you have a channel with some additional functions or packages, you simply add that channel with:

``` shell-session
$ sudo nix-channel --add https://example.com/my-nifty-channel foo
```

And you can access that channel using `<foo>`, for example in your configuration.nix:

``` nix
{
  imports = [ <foo/modules/shiny-little-module> ];
  environment.systemPackages = with import <foo/pkgs> {}; [ bar blah ];
  services.udev.extraRules = import <foo/lib/udev/mkrule.nix> {
    kernel = "eth*";
    attr.address = "00:1D:60:B9:6D:4F";
    name = "my_fast_network_card";
  };
}
```

Within nixpkgs, we shouldn't have `<nixos>` used anywhere anymore, so we shouldn't get into conflicts.

Cc: @edolstra

aszlig committed:
nixos: Add all of root's channels to NIX_PATH.

This is very useful if you want to distribute channels (and thus
expressions as well) in a similar fashion to Debians APT sources (or
PPAs or whatnot).

So, for example if you have a channel with some additional functions
or packages, you simply add that channel with:

sudo nix-channel --add https://example.com/my-nifty-channel foo

And you can access that channel using <foo>, for example in your
configuration.nix:

{
  imports = [ <foo/modules/shiny-little-module> ];
  environment.systemPackages = with import <foo/pkgs> {}; [ bar blah ];
  services.udev.extraRules = import <foo/lib/udev/mkrule.nix> {
    kernel = "eth*";
    attr.address = "00:1D:60:B9:6D:4F";
    name = "my_fast_network_card";
  };
}

Within nixpkgs, we shouldn't have <nixos> used anywhere anymore, so we
shouldn't get into conflicts.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos-rebuild: Update all channels of user root.

Should make it even easier to use custom channels, because whenever the
user does a "nixos-rebuild --upgrade", it will also upgrade possibly
used ("used" as in referenced in configuration.nix) channels besides
"nixos". And if you also ship a channel tied to a particular version of
nixpkgs or even remove the "nixos" channels, you won't run into
unexpected situations where the system is not updating your custom
channels.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

edolstra commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -93,9 +93,9 @@ if [ "$action" = switch -o "$action" = boot -o "$action" = test ]; then
+# If ‘--upgrade’ is given, run ‘nix-channel --update’.
 if [ -n "$upgrade" -a -z "$_NIXOS_REBUILD_REEXEC" ]; then
-    nix-channel --update nixos
+    nix-channel --update
```
Not sure about this. Users may not want all channels to be updated.

edolstra commented on nixos/modules/programs/environment.nix:
```diff
@@ -25,7 +25,8 @@ in
       { NIX_PATH =
-          [ "/nix/var/nix/profiles/per-user/root/channels/nixos"
+          [ "/nix/var/nix/profiles/per-user/root/channels"
+            "/nix/var/nix/profiles/per-user/root/channels/nixos"
```
A slight problem here is that you end up with an ambiguous/unintended situation if you add both the Nixpkgs and NixOS channel, because then NixOS will be updated from the Nixpkgs rather than NixOS channel.

aszlig committed:
nixos: Put root's channels to the end of NIX_PATH.

My original reason to put it at the beginning of NIX_PATH was to allow
shipping a particular version <nixpkgs> with a channel. But in order to
do that, we can still let the channel expression ship with a custom
version of nixpkgs by something like <channel/nixpkgs> and the builder
of the channel could also rewrite self-references.

So the inconvenience is now shifted towards the maintainer of the
channel rather than the user (which isn't nice, but better err on the
side of the developer rather than on the user), because as @edolstra
pointed out: Having the channels of root at the beginning of NIX_PATH
could have unintended side-effects if there a channel called nixpkgs.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos-rebuild: Conditionally update channels.

This partially reverts commit 3a4fd0bfc6fb34a496708ee95891acb1648eedde.

Addresses another concern by @edolstra that users might not want to
update *all* channels. We're now reverting to the old behaviour but
after updating the "nixos" channel, we just check whether the channel
ships with a file called ".update-on-nixos-rebuild" and if it exists, we
update that channel as well.

Other channels than these are not touched anymore.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
@edolstra: Thanks for reviewing, hope this looks better now :-)

domenkozar commented:
I really wish such PRs would include documentation. No one is going to fix that later :)

aszlig commented:
@domenkozar: Thanks for the heads-up. Right now this is just preparation and the main feature is not yet implemented in Hydra and release-tools. But I'll add an entry to release notes.

domenkozar commented:
:+1:
